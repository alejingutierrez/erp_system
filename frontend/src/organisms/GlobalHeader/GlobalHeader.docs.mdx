import { Meta, Canvas, Story, ArgsTable } from '@storybook/blocks';
import { GlobalHeader } from './GlobalHeader';

<Meta title="Organisms/GlobalHeader" of={GlobalHeader} />

# GlobalHeader

The `GlobalHeader` organizes primary navigation, search and user actions in a single top bar. It collapses the navigation items into a menu button on narrow screens.

<Canvas>
  <Story name="Desktop">
    <GlobalHeader
      logo="Fashion"
      title="ERP"
      navItems={[{ label: 'Dashboard', iconName: 'Home' }]}
      userName="Jane Doe"
      userMenuItems={[{ label: 'Logout' }]}
      notificationsCount={2}
      actionLabel="New"
      divider
    />
  </Story>
</Canvas>

<Canvas>
  <Story name="Mobile">
    <GlobalHeader
      logo="Fashion"
      title="ERP"
      navItems={[{ label: 'Dashboard', iconName: 'Home' }]}
      userName="Jane Doe"
      userMenuItems={[{ label: 'Logout' }]}
      notificationsCount={2}
      actionLabel="New"
    />
  </Story>
</Canvas>

<ArgsTable of={GlobalHeader} />

## Responsive behavior

Below the `sm` breakpoint the navigation items are hidden and can be accessed through a menu toggle. Use the `onNavigate`, `onSearch` and `onNotificationsOpen` callbacks to capture user actions.
