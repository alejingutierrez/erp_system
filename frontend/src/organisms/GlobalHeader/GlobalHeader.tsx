import * as React from 'react';
import { cva, type VariantProps } from 'class-variance-authority';

import { cn } from '@/lib/utils';
import { AppLayout } from '@/templates/AppLayout';
import { Sidebar } from '@/organisms/Sidebar';
import { TopBar } from '@/organisms/TopBar';
import type { NavItemProps } from '@/molecules/NavItem';
import type { DropdownMenuItem } from '@/molecules/DropdownMenu';
import { Level2Panel } from '../Level2Panel';

const globalHeaderVariants = cva('global-header', {
  variants: {},
  defaultVariants: {},
});

export interface GlobalHeaderProps
  extends React.HTMLAttributes<HTMLElement>,
    VariantProps<typeof globalHeaderVariants> {
  navItems?: NavItemProps[];
  actionLabel?: string;
  onAction?: () => void;
  notificationsCount?: number;
  userName?: string;
  userAvatarSrc?: string;
  userMenuItems?: DropdownMenuItem[];
  activePath?: string;
  onNavigate?: (path: string) => void;
  onSearch?: (term: string) => void;
  onNotificationsOpen?: () => void;
  level2Title?: string;
  level2Actions?: {
    label: string;
    onClick: () => void;
  }[];
}

export const GlobalHeader = React.forwardRef<HTMLElement, GlobalHeaderProps>(
  (
    {
      navItems = [],
      actionLabel,
      onAction,
      notificationsCount = 0,
      userName,
      userAvatarSrc,
      userMenuItems,
      onNavigate,
      onSearch,
      onNotificationsOpen,
      activePath,
      level2Title,
      level2Actions,
      className,
      children,
      ...props
    },
    ref,
  ) => {
    const [sidebarCollapsed, setSidebarCollapsed] = React.useState(false);
    const [activeLevel1, setActiveLevel1] = React.useState<string | null>(null);

    const handleNavigate = (path: string) => {
        setActiveLevel1(path);
        onNavigate?.(path);
    }

    return (
      <AppLayout
        {...props}
        ref={ref}
        className={cn(globalHeaderVariants(), className)}
        sidebar={
          <Sidebar
            navItems={navItems}
            collapsed={sidebarCollapsed}
            onNavigate={handleNavigate}
            activePath={activePath}
          />
        }
        topbar={
          <TopBar
            notificationsCount={notificationsCount}
            userName={userName}
            userAvatarSrc={userAvatarSrc}
            userMenuItems={userMenuItems}
            onSearch={onSearch}
            onNotificationsOpen={onNotificationsOpen}
            onToggleSidebar={() => setSidebarCollapsed(!sidebarCollapsed)}
          />
        }
      >
        {level2Title ? (
            <Level2Panel title={level2Title} actions={level2Actions}>
                {children}
            </Level2Panel>
        ) : children}
      </AppLayout>
    );
  },
);

GlobalHeader.displayName = 'GlobalHeader';
export { globalHeaderVariants };
