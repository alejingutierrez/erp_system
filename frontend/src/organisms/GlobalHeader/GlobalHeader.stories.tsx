import type { Meta, StoryObj } from '@storybook/react';
import { GlobalHeader, type GlobalHeaderProps } from './GlobalHeader';

const meta: Meta<GlobalHeaderProps> = {
  title: 'Organisms/GlobalHeader',
  component: GlobalHeader,
  tags: ['autodocs'],
  argTypes: {
    logo: { control: 'text' },
    title: { control: 'text' },
    navItems: { control: 'object' },
    actionLabel: { control: 'text' },
    notificationsCount: { control: 'number' },
    divider: { control: 'boolean' },
    onNavigate: { action: 'navigate', table: { category: 'Events' } },
    onSearch: { action: 'search', table: { category: 'Events' } },
    onNotificationsOpen: {
      action: 'notificationsOpen',
      table: { category: 'Events' },
    },
  },
};

export default meta;

type Story = StoryObj<typeof meta>;

const navItems = [
    { label: 'Panel', iconName: 'LayoutDashboard', path: '/panel' },
    { label: 'Productos & Catálogo', iconName: 'Package', path: '/productos' },
    { label: 'Inventario', iconName: 'Warehouse', path: '/inventario' },
    { label: 'Compras & Abastecimiento', iconName: 'ShoppingCart', path: '/compras' },
    { label: 'Ventas & Pedidos', iconName: 'ShoppingBag', path: '/ventas' },
    { label: 'CRM & Atención', iconName: 'Users', path: '/crm' },
    { label: 'Marketing & Campañas', iconName: 'Megaphone', path: '/marketing' },
    { label: 'CDP & Audiencias', iconName: 'Target', path: '/cdp' },
    { label: 'Logística & Fulfillment', iconName: 'Truck', path: '/logistica' },
    { label: 'Finanzas & Contabilidad', iconName: 'Landmark', path: '/finanzas' },
    { label: 'Analytics & BI', iconName: 'AreaChart', path: '/analytics' },
    { label: 'Integraciones', iconName: 'Plug', path: '/integraciones' },
    { label: 'Configuración', iconName: 'Settings', path: '/configuracion' },
    { label: 'Ayuda & Soporte', iconName: 'HelpCircle', path: '/ayuda' },
];

const storeMenuItems = [
    { label: 'Tienda Principal' },
    { label: 'Tienda Secundaria' },
];

export const Default: Story = {
  args: {
    navItems: navItems,
    notificationsCount: 3,
    userName: 'Jane Doe',
    userMenuItems: [
      { label: 'Profile' },
      { label: 'Logout' },
    ],
    storeMenuItems,
    activeStore: 'Tienda Principal',
    children: 'Contenido principal de la aplicación'
  },
};

export const WithLevel2Panel: Story = {
    args: {
        ...Default.args,
        activePath: '/productos',
        level2Title: 'Productos & Catálogo',
        level2Actions: [
            { label: 'Crear', onClick: () => alert('Crear') },
            { label: 'Importar', onClick: () => alert('Importar') },
            { label: 'Exportar', onClick: () => alert('Exportar') },
        ],
        children: 'Aquí iría el contenido del panel de nivel 2, como una tabla de productos.'
    }
}
