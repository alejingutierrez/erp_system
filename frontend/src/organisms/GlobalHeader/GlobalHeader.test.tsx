import { render, screen, fireEvent, act } from '@testing-library/react';
import { describe, it, expect, vi } from 'vitest';
import { GlobalHeader } from './GlobalHeader';

const items = [
  { label: 'Home', iconName: 'Home', path: '/home' },
  { label: 'Users', iconName: 'Users', path: '/users' },
];

const resize = (w: number) => {
  act(() => {
    window.innerWidth = w;
    window.dispatchEvent(new Event('resize'));
  });
};

describe('GlobalHeader', () => {
  it('renders logo and actions', () => {
    resize(1280);
    render(
      <GlobalHeader
        logo={<span>Logo</span>}
        navItems={items}
        userName="Ana"
        userMenuItems={[{ label: 'Logout' }]}
        actionLabel="New"
        notificationsCount={2}
      />,
    );
    expect(screen.getByText('Logo')).toBeInTheDocument();
    expect(screen.getByRole('search')).toBeInTheDocument();
    expect(screen.getByRole('button', { name: 'Notifications' })).toBeInTheDocument();
  });

  it('calls callbacks', () => {
    const onNav = vi.fn();
    const onSearch = vi.fn();
    const onNotif = vi.fn();
    render(
      <GlobalHeader
        navItems={items}
        onNavigate={onNav}
        onSearch={onSearch}
        onNotificationsOpen={onNotif}
        notificationsCount={1}
      />,
    );
    fireEvent.click(screen.getByText('Home'));
    expect(onNav).toHaveBeenCalledWith('/home');
    fireEvent.submit(screen.getByRole('search'));
    expect(onSearch).toHaveBeenCalled();
    fireEvent.click(screen.getByRole('button', { name: 'Notifications' }));
    expect(onNotif).toHaveBeenCalled();
  });

  it('collapses navigation on mobile', () => {
    resize(500);
    render(<GlobalHeader navItems={items} />);
    expect(screen.getByLabelText('Menu')).toBeInTheDocument();
    expect(screen.queryByText('Home')).not.toBeInTheDocument();
    fireEvent.click(screen.getByLabelText('Menu'));
    expect(screen.getByText('Home')).toBeInTheDocument();
    expect(screen.getByRole('search')).toBeInTheDocument();
    resize(1100);
    expect(screen.queryByLabelText('Menu')).not.toBeInTheDocument();
    expect(screen.getByText('Home')).toBeInTheDocument();
  });
});
