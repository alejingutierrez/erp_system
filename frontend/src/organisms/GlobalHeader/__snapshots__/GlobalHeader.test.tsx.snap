// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GlobalHeader > should render correctly 1`] = `
<div
  class="flex h-screen bg-background global-header"
>
  <aside
    class="flex flex-col gap-2 p-2 bg-background text-foreground transition-all duration-300 w-64"
  >
    <nav
      class="flex flex-col gap-2"
    >
      <button
        class="flex items-center gap-2 w-full rounded-md px-3 py-2 text-sm transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-secondary text-foreground hover:bg-muted"
        path="/panel"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-layout-dashboard inline-block align-middle w-5 h-5"
          fill="none"
          focusable="false"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="9"
            rx="1"
            width="7"
            x="3"
            y="3"
          />
          <rect
            height="5"
            rx="1"
            width="7"
            x="14"
            y="3"
          />
          <rect
            height="9"
            rx="1"
            width="7"
            x="14"
            y="12"
          />
          <rect
            height="5"
            rx="1"
            width="7"
            x="3"
            y="16"
          />
        </svg>
        <span
          class="font-sans text-lg font-normal flex-1 text-left"
        >
          Panel
        </span>
      </button>
      <button
        class="flex items-center gap-2 w-full rounded-md px-3 py-2 text-sm transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-secondary text-foreground hover:bg-muted"
        path="/productos"
        type="button"
      >
        <span
          class="font-sans text-lg font-normal flex-1 text-left"
        >
          Productos & Cat√°logo
        </span>
      </button>
    </nav>
  </aside>
  <div
    class="flex-1 flex flex-col overflow-hidden"
  >
    <header
      class="flex items-center gap-4 px-4 shadow-sm border-b border-border h-14 sm:h-16 glass backdrop-blur-md bg-background text-foreground"
    >
      <div
        class="flex items-center gap-2"
      >
        <button
          aria-label="Toggle Sidebar"
          class="group inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-all duration-300 ease-in-out focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 focus-visible:ring-secondary disabled:pointer-events-none disabled:opacity-50 hover:-translate-y-px hover:shadow-sm active:translate-y-0 active:shadow-md active:scale-105 p-0 border-transparent h-9 w-9 bg-primary text-primary-foreground hover:bg-primary/90"
        />
        <div
          class="flex-1"
        >
          <form
            aria-label="Buscar"
            class="search-bar flex items-center gap-2 w-80 max-w-xs"
            role="search"
          >
            <div
              class="relative group"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-search inline-block align-middle absolute left-3 top-1/2 -translate-y-1/2 text-muted"
                fill="none"
                focusable="false"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m21 21-4.34-4.34"
                />
                <circle
                  cx="11"
                  cy="11"
                  r="8"
                />
              </svg>
              <input
                class="peer block w-full rounded-md border bg-white text-foreground placeholder:text-muted-foreground focus:outline-none focus:ring-2 disabled:cursor-not-allowed disabled:opacity-50 py-2 px-3 text-base pl-10 border-primary focus:ring-primary flex-1"
                placeholder="Buscar..."
                type="search"
                value=""
              />
            </div>
          </form>
        </div>
      </div>
      <div
        class="flex items-center gap-2"
      >
        <button
          aria-label="Buscar"
          class="group inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-all duration-300 ease-in-out focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 focus-visible:ring-secondary disabled:pointer-events-none disabled:opacity-50 hover:-translate-y-px hover:shadow-sm active:translate-y-0 active:shadow-md active:scale-105 p-0 border-transparent h-9 w-9 bg-primary text-primary-foreground hover:bg-primary/90 sm:hidden"
        />
        <button
          aria-label="Notifications"
          class="relative inline-flex"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-bell inline-block align-middle w-5 h-5 text-primary"
            fill="none"
            focusable="false"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10.268 21a2 2 0 0 0 3.464 0"
            />
            <path
              d="M3.262 15.326A1 1 0 0 0 4 17h16a1 1 0 0 0 .74-1.673C19.41 13.956 18 12.499 18 8A6 6 0 0 0 6 8c0 4.499-1.411 5.956-2.738 7.326"
            />
          </svg>
        </button>
        <button
          aria-label="Toggle theme"
          class="group inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-all duration-300 ease-in-out focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 focus-visible:ring-secondary disabled:pointer-events-none disabled:opacity-50 hover:-translate-y-px hover:shadow-sm active:translate-y-0 active:shadow-md active:scale-105 p-0 border-transparent h-9 w-9 bg-primary text-primary-foreground hover:bg-primary/90"
        />
        <div
          class="flex items-center gap-2"
        >
          <div
            class="relative inline-flex items-center justify-center overflow-hidden rounded-full select-none shadow-sm transition-all hover:shadow-md hover:-translate-y-0.5 h-6 w-6 text-xs bg-muted text-muted-foreground"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-user h-2/3 w-2/3"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2"
              />
              <circle
                cx="12"
                cy="7"
                r="4"
              />
            </svg>
          </div>
          <span
            class="text-sm font-medium hidden sm:inline"
          />
        </div>
      </div>
    </header>
    <main
      class="flex-1 overflow-x-hidden overflow-y-auto bg-background"
    />
  </div>
</div>
`;
